### PR TITLE
 Add Auth interoperability library.

### PR DESCRIPTION
**Pull request by ryanwilson**
This allows other SDKs to retrieve the user ID and fetch a new
Auth token in a type-safe way through the Core container.

### CODE REVIEW

File: Interoperability/Auth/Public/FirebaseAuthInterop.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright 2017 Google
```
Comment (paulb777) [inline Interoperability/Auth/Public/FirebaseAuthInterop.h]: 2018

File: Interoperability/Auth/Public/FirebaseAuthInterop.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
+/*
```
Comment (paulb777) [inline Interoperability/Auth/Public/FirebaseAuthInterop.h]: What's the purpose of this file?  I believe we only construct umbrella headers manually for legacy compatibility reasons.

File: FirebaseAuthInteroperable.podspec [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  # on each other.
+  s.source           = {
+    :git => 'https://github.com/firebase/firebase-ios-sdk.git',
+    :tag => s.version.to_s
+  }
+  s.social_media_url = 'https://twitter.com/Firebase'
+  s.ios.deployment_target = '8.0'
+  s.osx.deployment_target = '10.10'
+  s.tvos.deployment_target = '10.0'
+  s.preserve_paths = 'README.md'
```
Comment (paulb777) [inline FirebaseAuthInteroperable.podspec]: Delete? Or why is this here?

File: FirebaseAuthInteroperable.podspec [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  s.source_files = 'Interoperability/Auth/**/*.h'
+  s.public_header_files = 'Interoperability/Auth/Public/*.h'
```
[the remark this answers is not in the document]
Comment (paulb777) [reply FirebaseAuthInteroperable.podspec]: Hmm. I suppose there's two approaches:

1. Allow major changes to the Interoperability API's but not expose them transitively through their dependents.
2. Don't have any public APIs in the Interoperability pod.

I have a preference for 2 since we don't want these to be considered public, but I could be convinced otherwise.

**Comment by paulb777** [comment]
Also, looks like it needs a rebase and green travis run.

File: FirebaseAuthInteroperable.podspec [hunk 5]
```diff
@@ -23,5 +23,5 @@ Pod::Spec.new do |s|
   s.osx.deployment_target = '10.10'
   s.tvos.deployment_target = '10.0'
   s.source_files = 'Interoperability/Auth/**/*.h'
-  s.public_header_files = 'Interoperability/Auth/Public/*.h'
```
Comment (paulb777) [inline FirebaseAuthInteroperable.podspec]: I believe you also need:

`s.public_header_files = 'Interoperability/Auth/Private/*.h'`

File: Firebase/Auth/Source/FIRAuth.m [hunk 6]
```diff
@@ -18,6 +18,10 @@
 
 #import "FIRAuth_Internal.h"
 
+#if __has_include(<UIKit/UIKit.h>)
```
Comment (paulb777) [inline Firebase/Auth/Source/FIRAuth.m]: Nice! For consistency, consider using the same #if for the UIKit usage below.

Comment (ryanwilson) [reply Firebase/Auth/Source/FIRAuth.m]: Thanks! We should look at moving others over to it.

So I was considering it but I think those are probably better off with specific targets just in case since the APIs can differ (I'm fairly certain some things in UIKit are available on iOS but not tvOS, so that check may not be exactly correct).

**Comment by ryanwilson** [comment]
We moved the headers back into public since CocoaPods will skip over the framework with no public header and no `.m` files during the `.xcconfig` file generation process, among other issues. Moving the header back into the public headers fixes this.
